persistence/state.go

```diff
@@ -0,0 +1,187 @@
+package persistence
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"log"
+
+	"github.com/celestiaorg/smt"
+	"github.com/pokt-network/pocket/persistence/types"
+	"google.golang.org/protobuf/proto"
+)
+
+type MerkleTree float64
+
+// A work-in-progress list of all the trees we need to update to maintain the overall state
+const (
+	// Actor  Merkle Trees
+	appMerkleTree MerkleTree = iota
+	valMerkleTree
+	fishMerkleTree
+	serviceNodeMerkleTree
+
+	// Account Merkle Trees
+	accountMerkleTree
+	poolMerkleTree
+
+	// Data / State Merkle Trees
+	blocksMerkleTree
+	paramsMerkleTree
+	flagsMerkleTree
+
+	// Used for iteration purposes only - see https://stackoverflow.com/a/64178235/768439
+	lastMerkleTree
+)
+
+var actorTypeToMerkleTreeName map[types.ActorType]MerkleTree = map[types.ActorType]MerkleTree{
+	types.ActorType_App:  appMerkleTree,
+	types.ActorType_Val:  valMerkleTree,
+	types.ActorType_Fish: fishMerkleTree,
+	types.ActorType_Node: serviceNodeMerkleTree,
+}
+
+var merkleTreeToActorTypeName map[MerkleTree]types.ActorType = map[MerkleTree]types.ActorType{
+	appMerkleTree:         types.ActorType_App,
+	valMerkleTree:         types.ActorType_Val,
+	fishMerkleTree:        types.ActorType_Fish,
+	serviceNodeMerkleTree: types.ActorType_Node,
+}
+
+var actorTypeToSchemaName map[types.ActorType]types.ProtocolActorSchema = map[types.ActorType]types.ProtocolActorSchema{
+	types.ActorType_App:  types.ApplicationActor,
+	types.ActorType_Val:  types.ValidatorActor,
+	types.ActorType_Fish: types.FishermanActor,
+	types.ActorType_Node: types.ServiceNodeActor,
+}
+
+func newMerkleTrees() (map[MerkleTree]*smt.SparseMerkleTree, error) {
+	// We need a separate Merkle tree for each type of actor or storage
+	trees := make(map[MerkleTree]*smt.SparseMerkleTree, int(lastMerkleTree))
+
+	for treeType := MerkleTree(0); treeType < lastMerkleTree; treeType++ {
+		// TODO_IN_THIS_COMMIT: Rather than using `NewSimpleMap`, use a disk based key-value store
+		nodeStore := smt.NewSimpleMap()
+		valueStore := smt.NewSimpleMap()
+
+		trees[treeType] = smt.NewSparseMerkleTree(nodeStore, valueStore, sha256.New())
+	}
+	return trees, nil
+}
+
+func loadMerkleTrees(map[MerkleTree]*smt.SparseMerkleTree, error) {
+	log.Fatalf("loadMerkleTrees not implemented yet")
+}
+
+func (p *PostgresContext) updateStateHash() error {
+	// Update all the merkle trees
+	for treeType := MerkleTree(0); treeType < lastMerkleTree; treeType++ {
```

Comment:
The pattern of a `for loop` with a `switch` statement is not my favorite.

It makes me think there might be an opportunity to use a map like structure `merkleTrees[treeType]`.

Reply:
I try a few different approaches but did not succeed. Mainly because the data (e.g. blocks), accounts (pools, accounts) and actors all have slightly different logic so it ended up making the code more complex than not.

Resolving for now, but will keep thinking.